cmd/olm/main.go

```diff
@@ -257,10 +258,30 @@ func main() {
 		go monitor.Run(op.Done())
 	}
 
+	// Emit CSV metric
+	if err = ensureCSVMetric(logger, crClient); err != nil {
+		logger.WithError(err).Fatalf("error emitting CSV metric")
+	}
+
 	// Start the controller manager
 	if err := mgr.Start(ctx); err != nil {
 		logger.WithError(err).Fatal("controller manager stopped")
 	}
 
 	<-op.Done()
 }
+
+func ensureCSVMetric(logger *logrus.Logger, c *versioned.Clientset) error {
```

Comment:
This logic looks perfect, although I'm not sure if the main function is the write place for it. Is there a reason including this logic when a new olm controller is created instead might be a bad idea?

Reply:
The informer cache is not yet populated at this moment - I see that as an alternative place for this call. Just when the cache gets populated.

Reply:
Nothing immediate comes to mind as to why this implementation may be a bad idea, but maybe others have stronger opinions on this - we could also wrap this in a Runnable instance to abstract away some of the prerequisites before starting the controller/informers/etc. if we're worried about control flow in the driver function.

Reply:
I guess one main downsides is that performing a list call on a non-cached client may be inefficient in the context of a cluster with a significant amount of CSVs.

Reply:
I think I was wrong assuming that the List call is from a cold cache.
[This](https://github.com/operator-framework/operator-lifecycle-manager/blob/6a52c02a42d8623a369e6501520919be066d1644/cmd/olm/main.go#L210) is where the main function waits for the cache sync and [this](https://github.com/operator-framework/operator-lifecycle-manager/blob/6a52c02a42d8623a369e6501520919be066d1644/pkg/lib/queueinformer/queueinformer_operator.go#L182) is the deferred call that signals that the cache has synced.

ad Placing this call to main.go: What is the alternative? The way I look at it is that it shouldn't happen before the informer cache has synced and not after the controllers have started.

Reply:
> not after the controllers have started

Could you elaborate a little more on why not?

Reply:
1) Then the controller could emit the metric during CSV reconciliation
2) It prolongs the period when the metric is not accurately set - thus we don't target the core problem

Reply:
https://github.com/operator-framework/operator-lifecycle-manager/blob/6a52c02a42d8623a369e6501520919be066d1644/pkg/lib/queueinformer/queueinformer_operator.go#L213-L218

couldn't it go in between here? after caches have synced but before controllers have started?

Reply:
That looks to be a fairly generic function. [Catalog operator](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/cmd/catalog/main.go#L139)(and other clients of that lib) would then be exposing CSV metric.

[OLM operator](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/olm/operator.go#L61) doesn't have a sense when the cache has been synced. But it could be done - we could pass the ready channel to NewOperator() and spin up a goroutine that would emit the metric once the cache is synced.
It sounds quite complicated, but the emit logic would be in the operator pkg.